File: admin/src/api/erp_inventory_transfer.ts
import { PaginatedRequest, PaginatedResponse } from '@/base/page';
import { api } from '@/utils/request';
import { ErpInventoryTransferDetailBaseResponse, ErpInventoryTransferDetailRequest } from './erp_inventory_transfer_detail';
import { ErpInventoryTransferAttachmentBaseResponse, ErpInventoryTransferAttachmentRequest } from './erp_inventory_transfer_attachment';

const apis = {
  create: '/erp/erp_inventory_transfer/create', // 新增
  update: '/erp/erp_inventory_transfer/update', // 修改
  delete: '/erp/erp_inventory_transfer/delete', // 删除
  get: '/erp/erp_inventory_transfer/get', // 单条查询
  get_base: '/erp/erp_inventory_transfer/get_base', // 单条查询
  list: '/erp/erp_inventory_transfer/list', // 列表查询
  page: '/erp/erp_inventory_transfer/page', // 分页查询
}

export interface ErpInventoryTransferRequest {
  id?: number; // 调拨记录ID
  transfer_date: string; // 调拨日期
  remarks: string; // 备注

  details: ErpInventoryTransferDetailRequest[]; // 调拨列表
  attachments: ErpInventoryTransferAttachmentRequest[]; // 调拨附件列表
}

export interface ErpInventoryTransferResponse {
  id: number; // 调拨记录ID
  order_number: number; // 订单编号
  transfer_date: string; // 调拨日期
  remarks: string; // 备注
  department_code: string; // 部门编码
  department_id: number; // 部门ID
  creator: number; // 创建者ID
  create_time: string; // 创建时间
  updater: number; // 更新者ID
  update_time: string; // 更新时间
}

export interface ErpInventoryTransferBaseResponse {
  id: number; // 调拨记录ID
  order_number: number; // 订单编号
  transfer_date: string; // 调拨日期
  remarks: string; // 备注

  details: ErpInventoryTransferDetailBaseResponse[]; // 调拨列表
  attachments: ErpInventoryTransferAttachmentBaseResponse[]; // 调拨附件列表
}

export interface ErpInventoryTransferQueryCondition extends PaginatedRequest {

}

export const createErpInventoryTransfer = (erp_inventory_transfer: ErpInventoryTransferRequest): Promise<number> => {
  return api.post<number>(apis.create, erp_inventory_transfer);
}

export const updateErpInventoryTransfer = (erp_inventory_transfer: ErpInventoryTransferRequest): Promise<void> => {
  return api.post<void>(apis.update, erp_inventory_transfer);
}

export const deleteErpInventoryTransfer = (id: number): Promise<void> => {
  return api.post<void>(`${apis.delete}/${id}`);
}

export const getErpInventoryTransfer = (id: number): Promise<ErpInventoryTransferResponse> => {
  return api.get<ErpInventoryTransferResponse>(`${apis.get}/${id}`);
}

export const getBaseErpInventoryTransfer = (id: number): Promise<ErpInventoryTransferBaseResponse> => {
  return api.get<ErpInventoryTransferBaseResponse>(`${apis.get_base}/${id}`);
}

export const listErpInventoryTransfer = (): Promise<Array<ErpInventoryTransferResponse>> => {
  return api.get<Array<ErpInventoryTransferResponse>>(apis.list);
}

export const pageErpInventoryTransfer = (condition: ErpInventoryTransferQueryCondition): Promise<PaginatedResponse<ErpInventoryTransferResponse>> => {
  return api.get<PaginatedResponse<ErpInventoryTransferResponse>>(apis.page, condition);
}

File: server/module/module-erp/erp-server/src/service/erp_inventory_transfer.rs
use common::interceptor::orm::simple_support::SimpleSupport;
use common::utils::snowflake_generator::SnowflakeGenerator;
use sea_orm::{ActiveModelTrait, ColumnTrait, Condition, DatabaseConnection, EntityTrait, Order, PaginatorTrait, QueryFilter, QueryOrder, TransactionTrait};
use crate::model::erp_inventory_transfer::{Model as ErpInventoryTransferModel, ActiveModel as ErpInventoryTransferActiveModel, Entity as ErpInventoryTransferEntity, Column};
use crate::service::{erp_inventory_transfer_attachment, erp_inventory_transfer_detail};
use erp_model::request::erp_inventory_transfer::{CreateErpInventoryTransferRequest, UpdateErpInventoryTransferRequest, PaginatedKeywordRequest};
use erp_model::response::erp_inventory_transfer::{ErpInventoryTransferBaseResponse, ErpInventoryTransferResponse};
use crate::convert::erp_inventory_transfer::{create_request_to_model, model_to_base_response, model_to_response, update_request_to_model};
use anyhow::{anyhow, Context, Result};
use sea_orm::ActiveValue::Set;
use common::constants::enum_constants::{STATUS_DISABLE, STATUS_ENABLE};
use common::base::page::PaginatedResponse;
use common::context::context::LoginUserContext;
use common::interceptor::orm::active_filter::ActiveFilterEntityTrait;

pub async fn create(db: &DatabaseConnection, login_user: LoginUserContext, request: CreateErpInventoryTransferRequest) -> Result<i64> {
    let mut erp_inventory_transfer = create_request_to_model(&request);
    // 生成订单编号
    let generator = SnowflakeGenerator::new();
    match generator.generate() {
        Ok(id) => erp_inventory_transfer.order_number = Set(id),
        Err(e) => return Err(anyhow!("订单编号生成失败")),
    }
    erp_inventory_transfer.department_id = Set(login_user.department_id.clone());
    erp_inventory_transfer.department_code = Set(login_user.department_code.clone());
    erp_inventory_transfer.creator = Set(Some(login_user.id.clone()));
    erp_inventory_transfer.updater = Set(Some(login_user.id.clone()));
    erp_inventory_transfer.tenant_id = Set(login_user.tenant_id.clone());

    // 开启事务
    let txn = db.begin().await?;
    // 创建订单
    let erp_inventory_transfer = erp_inventory_transfer.insert(db).await?;
    // 创建订单详情
    erp_inventory_transfer_detail::create_batch(&db, &txn, login_user.clone(), erp_inventory_transfer.clone(), request.details).await?;
    // 创建订单文件
    erp_inventory_transfer_attachment::create_batch(&db, &txn, login_user.clone(), erp_inventory_transfer.id, request.attachments).await?;
    // 提交事务
    txn.commit().await.with_context(|| "Failed to commit transaction")?;
    Ok(erp_inventory_transfer.id)
}

pub async fn update(db: &DatabaseConnection, login_user: LoginUserContext, request: UpdateErpInventoryTransferRequest) -> Result<()> {
    let erp_inventory_transfer = ErpInventoryTransferEntity::find_by_id(request.id)
        .filter(Column::TenantId.eq(login_user.tenant_id))
        .one(db)
        .await?
        .ok_or_else(|| anyhow!("记录未找到"))?;

    let inventory_transfer = erp_inventory_transfer.clone();

    let mut erp_inventory_transfer = update_request_to_model(&request, erp_inventory_transfer);
    erp_inventory_transfer.updater = Set(Some(login_user.id));
    // 开启事务
    let txn = db.begin().await?;
    // 修改订单
    erp_inventory_transfer.update(db).await?;
    // 修改订单详情
    erp_inventory_transfer_detail::update_batch(&db, &txn, login_user.clone(), inventory_transfer.clone(), request.details).await?;
    // 修改订单文件
    erp_inventory_transfer_attachment::update_batch(&db, &txn, login_user.clone(), inventory_transfer, request.attachments).await?;
    // 提交事务
    txn.commit().await.with_context(|| "Failed to commit transaction")?;
    Ok(())
}

pub async fn delete(db: &DatabaseConnection, login_user: LoginUserContext, id: i64) -> Result<()> {
    let erp_inventory_transfer = ErpInventoryTransferActiveModel {
        id: Set(id),
        updater: Set(Some(login_user.id)),
        deleted: Set(true),
        ..Default::default()
    };
    erp_inventory_transfer.update(db).await?;
    Ok(())
}

pub async fn get_by_id(db: &DatabaseConnection, login_user: LoginUserContext, id: i64) -> Result<Option<ErpInventoryTransferResponse>> {
    let condition = Condition::all()
            .add(Column::Id.eq(id))
            .add(Column::TenantId.eq(login_user.tenant_id));
            
    let erp_inventory_transfer = ErpInventoryTransferEntity::find_active_with_condition(condition)
        .one(db).await?;
    Ok(erp_inventory_transfer.map(model_to_response))
}

pub async fn get_base_by_id(db: &DatabaseConnection, login_user: LoginUserContext, id: i64) -> Result<Option<ErpInventoryTransferBaseResponse>> {
    let condition = Condition::all()
            .add(Column::Id.eq(id))
            .add(Column::TenantId.eq(login_user.tenant_id));
            
    let erp_inventory_transfer = ErpInventoryTransferEntity::find_active_with_data_permission(login_user.clone())
        .filter(condition)
        .one(db).await?;
    
    if erp_inventory_transfer.is_none() {
        return Ok(None);
    }
    let erp_inventory_transfer = erp_inventory_transfer.unwrap();
    let details = erp_inventory_transfer_detail::list_by_order_id(&db, login_user.clone(), id).await?;
    let attachments = erp_inventory_transfer_attachment::list_by_order_id(&db, login_user, id).await?;
    Ok(Some(model_to_base_response(erp_inventory_transfer, details, attachments)))
}

pub async fn get_paginated(db: &DatabaseConnection, login_user: LoginUserContext, params: PaginatedKeywordRequest) -> Result<PaginatedResponse<ErpInventoryTransferResponse>> {
    let condition = Condition::all().add(Column::TenantId.eq(login_user.tenant_id));let paginator = ErpInventoryTransferEntity::find_active_with_condition(condition)
        .support_filter(params.base.filter_field, params.base.filter_operator, params.base.filter_value)
        .support_order(params.base.sort_field, params.base.sort, Some(vec![(Column::Id, Order::Asc)]))
        .paginate(db, params.base.size);

    let total = paginator.num_items().await?;
    let total_pages = (total + params.base.size - 1) / params.base.size; // 向上取整
    let list = paginator
        .fetch_page(params.base.page - 1) // SeaORM 页码从 0 开始，所以减 1
        .await?
        .into_iter()
        .map(model_to_response)
        .collect();

    Ok(PaginatedResponse {
        list,
        total_pages,
        page: params.base.page,
        size: params.base.size,
        total,
    })
}

pub async fn list(db: &DatabaseConnection, login_user: LoginUserContext) -> Result<Vec<ErpInventoryTransferResponse>> {
    let condition = Condition::all().add(Column::TenantId.eq(login_user.tenant_id));let list = ErpInventoryTransferEntity::find_active_with_condition(condition)
        .all(db).await?;
    Ok(list.into_iter().map(model_to_response).collect())
}
